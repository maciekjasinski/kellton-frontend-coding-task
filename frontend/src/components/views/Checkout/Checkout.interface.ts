export interface FormikValuesInterface {
  name: string;
  surname: string;
  phoneNumber: string;
  email: string;
  dateOfBirth: string;
  address: string;
  city: string;
  state: string;
  zipCode: string;
}
