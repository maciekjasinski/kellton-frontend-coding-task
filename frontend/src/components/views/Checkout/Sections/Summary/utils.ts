import { MinifigPartType } from '@/types/minifig';
import axios from 'axios';

export const fetchParts = async (set_id: string) => {
  return await axios
    .get(
      `https://rebrickable.com/api/v3/lego/minifigs/${set_id}/parts/?key=${import.meta.env.VITE_REBRICKABLE_API_KEY}`,
    )
    .then((res) => {
      return res.data.results.map((each: any) => {
        const { part } = each;
        return {
          id: part.part_num,
          name: part.name,
          picture: part.part_img_url || null,
        };
      }) as MinifigPartType[];
    }).catch(() => {
      return [];
    });
};
