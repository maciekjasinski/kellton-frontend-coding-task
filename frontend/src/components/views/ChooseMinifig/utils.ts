import { MinifigType } from '@/types/minifig';
import axios from 'axios';

const getRandomId = (arrayLength: number) => {
  const uniqueIds = new Set<number>();

  while (uniqueIds.size < 3) {
    const id = Math.floor(Math.random() * arrayLength);
    uniqueIds.add(id);
  }

  return Array.from(uniqueIds);
}

export const fetchMinifigs = async () => {
  return await axios
    .get(`https://rebrickable.com/api/v3/lego/minifigs/?key=${import.meta.env.VITE_REBRICKABLE_API_KEY}&page_size=1000&in_theme_id=${import.meta.env.VITE_HARRY_POTTER_THEME_ID}`)
    .then((res) => {
      const randomIds = getRandomId(res.data.count);
      const randomMinifigs = randomIds.map((id: number) => res.data.results[id]);
      return randomMinifigs.map(
        (each: { name: string; set_img_url: string | null; set_url: string; set_num: string }) => {
          return {
            name: each.name,
            picture: each.set_img_url || null,
            detailsUrl: each.set_url,
            setId: each.set_num,
          };
        },
      ) as MinifigType[];
    }).catch(() => {
      return [];
    })
};
