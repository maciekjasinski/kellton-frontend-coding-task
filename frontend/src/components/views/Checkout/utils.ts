import * as yup from 'yup';
import { parsePhoneNumber } from 'libphonenumber-js';
import { EMAIL_REGEX } from '@/utils';

export const initialValues = {
  name: '',
  surname: '',
  phoneNumber: '',
  email: '',
  dateOfBirth: '',
  address: '',
  city: '',
  state: '',
  zipCode: '',
};

export const validationSchema = yup.object().shape({
  name: yup.string().required('Name is required'),
  surname: yup.string().required('Surname is required'),
  phoneNumber: yup
    .string()
    .test('phoneNumber', 'Phone number is not valid', function () {
      const fieldValue = this.parent.phoneNumber;
      if (fieldValue) {
        try {
          const phoneNumner = parsePhoneNumber(fieldValue, 'US');
          const isValid = phoneNumner.isValid();
          const isPossible = phoneNumner.isPossible();
          return isValid && isPossible;
        } catch (err) {
          return this.createError();
        }
      }
      return this.createError();
    })
    .required('Phone number is required'),
  email: yup.string().required('Email is required').matches(EMAIL_REGEX, 'Invalid email'),
  dateOfBirth: yup
    .date()
    .required('Date of birth is required')
    .test('dateOfBirth', 'Should be greater than 18', function () {
      const value = this.parent.dateOfBirth;
      const dateOfBirth = new Date(value);
      const nowDate = new Date();
      const valid = nowDate.getFullYear() - dateOfBirth.getFullYear() >= 18;
      return !valid ? this.createError() : valid;
    }),
  address: yup.string().required('Address is required'),
  city: yup.string().required('City is required'),
  state: yup.string().required('State is required'),
  zipCode: yup.string().required('Zip code is required'),
});
