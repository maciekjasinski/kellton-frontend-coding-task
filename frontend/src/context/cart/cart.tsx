import { ReactNode, useReducer } from 'react';
import { CartStateContext, initialValues } from './utils';
import { cartReducer } from './reducer';

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer(cartReducer, initialValues);

  const value = {
    ...state,
    cartDispatch: dispatch,
  };

  return <CartStateContext.Provider value={value}>{children}</CartStateContext.Provider>;
};
