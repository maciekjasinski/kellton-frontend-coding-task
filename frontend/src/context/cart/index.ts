import { useContext } from 'react';

import { CartProvider } from './cart';
import { CartStateContext } from './utils';

const useCart = () => {
  const context = useContext(CartStateContext);
  if (context === undefined) {
    throw new Error('useCart must be used within an CartProvider');
  }
  return context;
};

export { CartProvider, useCart };
