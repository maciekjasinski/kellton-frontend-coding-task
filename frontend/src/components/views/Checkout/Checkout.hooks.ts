import { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import axios from 'axios';
import { useCart } from '@/context/cart';
import { FormikValuesInterface } from './Checkout.interface';

export const useCheckout = () => {
  const navigate = useNavigate();
  const { selectedMinifig, cartDispatch } = useCart();

  const handleSubmit = (values: FormikValuesInterface) => {
    return axios
      .post('/api/checkout', values)
      .then((res) => {
        cartDispatch({ type: 'clearCart' });
        navigate('/');
      })
      .catch((err) => console.log(err));
  };

  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);

  return {
    selectedMinifig,
    handleSubmit,
  };
};
