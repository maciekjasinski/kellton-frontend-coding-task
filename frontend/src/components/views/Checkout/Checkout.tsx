import { Link } from 'react-router-dom';
import { Formik, Form } from 'formik';
import { Form as FromSection } from './Sections/Form';
import { Summary } from './Sections/Summary';
import { useCheckout } from './Checkout.hooks';
import { initialValues, validationSchema } from './utils';

export const Checkout = () => {
  const { selectedMinifig, handleSubmit } = useCheckout();

  if (!selectedMinifig) {
    return (
      <div className="flex min-h-dvh items-center justify-center p-4 xl:mx-auto xl:w-maxWidth">
        <div className="flex w-full flex-col items-center gap-4 rounded-lg bg-white p-4 lg:w-fit lg:p-6">
          <h1 className="text-center text-2xl font-bold uppercase text-info-content">Your cart is empty!</h1>
          <p className="text-center text-base text-gray-400">Please choose your minifig.</p>
          <Link to="/choose-minifig" className="w-full lg:w-fit">
            <button className="btn btn-primary w-full text-white">Select minifig</button>
          </Link>
        </div>
      </div>
    );
  }

  return (
    <div className="flex flex-col gap-6 p-4 xl:mx-auto xl:min-h-dvh xl:w-maxWidth xl:justify-center xl:p-0 xl:py-12">
      <h1 className="text-2xl font-bold uppercase text-white">Shipping details</h1>
      <Formik initialValues={initialValues} onSubmit={handleSubmit} validationSchema={validationSchema}>
        <Form className="flex flex-col gap-16 lg:grid lg:grid-cols-checkoutGrid lg:gap-24">
          <FromSection />
          <Summary />
        </Form>
      </Formik>
    </div>
  );
};
