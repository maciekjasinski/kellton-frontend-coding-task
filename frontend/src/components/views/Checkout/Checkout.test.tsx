import { expect, describe, it, vi } from 'vitest';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { CartStateContext } from '@/context/cart/utils';
import { Checkout } from './Checkout';
import { validationSchema } from './utils';

// @ts-ignore
window.scrollTo = vi.fn();

describe('ChooseMinifig', () => {
  it('should create a snapshot with empty cart', () => {
    const dispatch = vi.fn();
    const page = renderer.create(
      <BrowserRouter>
        <CartStateContext.Provider
          value={{
            cartDispatch: dispatch,
            selectedMinifig: null,
          }}
        >
          <Checkout />
        </CartStateContext.Provider>
      </BrowserRouter>,
    );
    expect(page).toMatchSnapshot();
  });

  it('should create a snapshot', () => {
    const dispatch = vi.fn();
    const page = renderer.create(
      <BrowserRouter>
        <CartStateContext.Provider
          value={{
            cartDispatch: dispatch,
            selectedMinifig: {
              name: 'Minifig',
              detailsUrl: '',
              picture: null,
              setId: 's1',
              parts: [
                {
                  id: 'p1',
                  name: 'Part 1',
                  picture: null,
                },
                {
                  id: 'p2',
                  name: 'Part 2',
                  picture: null,
                },
              ],
            },
          }}
        >
          <Checkout />
        </CartStateContext.Provider>
      </BrowserRouter>,
    );
    expect(page).toMatchSnapshot();
  });
});

describe('Validation', () => {
  it('should validate form successfully', async () => {
    const isValid = await validationSchema.isValid({
      name: 'Joe',
      surname: 'Doe',
      phoneNumber: '12312312312',
      email: 'joe@doe.com',
      dateOfBirth: '2000-03-29',
      address: 'Street 1',
      city: 'New Yrok',
      state: 'New York',
      zipCode: '10000',
    });
    expect(isValid).toEqual(true);
  });

  it('should throw error with empty form', async () => {
    const isValid = await validationSchema.isValid({
      name: '',
      surname: '',
      phoneNumber: '',
      email: '',
      dateOfBirth: '',
      address: '',
      city: '',
      state: '',
      zipCode: '',
    });
    expect(isValid).toEqual(false);
  });

  it('phone number should be not valid', async () => {
    const isValid = await validationSchema.isValid({
      name: 'Joe',
      surname: 'Doe',
      phoneNumber: '123',
      email: 'joe@doe.com',
      dateOfBirth: '2000-03-29',
      address: 'Street 1',
      city: 'New Yrok',
      state: 'New York',
      zipCode: '10000',
    });
    expect(isValid).toEqual(false);
  });

  it('date of birth should be minimum 18 years old', async () => {
    const isValid = await validationSchema.isValid({
      name: 'Joe',
      surname: 'Doe',
      phoneNumber: '12312312312',
      email: 'joe@doe.com',
      dateOfBirth: '2007-03-29',
      address: 'Street 1',
      city: 'New Yrok',
      state: 'New York',
      zipCode: '10000',
    });
    expect(isValid).toEqual(false);
  });
});
