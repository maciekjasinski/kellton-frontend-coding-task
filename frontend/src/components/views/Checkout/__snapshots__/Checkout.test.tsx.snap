// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChooseMinifig > should create a snapshot 1`] = `
<div
  className="flex flex-col gap-6 p-4 xl:mx-auto xl:min-h-dvh xl:w-maxWidth xl:p-0 xl:py-12"
>
  <form
    action="#"
    className="flex flex-col gap-16 lg:grid lg:min-h-[calc(100dvh-96px)] lg:grid-cols-checkoutGrid lg:gap-24"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="flex flex-col gap-8 lg:justify-center"
    >
      <h1
        className="text-2xl font-extrabold uppercase text-white lg:text-3xl"
      >
        Shipping details
      </h1>
      <div
        className="flex flex-col gap-4"
      >
        <div
          className="flex flex-col gap-4 lg:w-full lg:flex-row"
        >
          <label
            className="lg:min-w-xs form-control w-full"
          >
            <div
              className="label px-0 pb-1"
            >
              <span
                className="label-text font-bold text-white"
              >
                Name
              </span>
            </div>
            <input
              className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Type here"
              type="text"
              value=""
            />
          </label>
          <label
            className="lg:min-w-xs form-control w-full"
          >
            <div
              className="label px-0 pb-1"
            >
              <span
                className="label-text font-bold text-white"
              >
                Surname
              </span>
            </div>
            <input
              className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
              name="surname"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Type here"
              type="text"
              value=""
            />
          </label>
        </div>
        <label
          className="lg:min-w-xs form-control w-full"
        >
          <div
            className="label px-0 pb-1"
          >
            <span
              className="label-text font-bold text-white"
            >
              Phone number
            </span>
          </div>
          <input
            className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
            name="phoneNumber"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type here"
            type="text"
            value=""
          />
        </label>
        <label
          className="lg:min-w-xs form-control w-full"
        >
          <div
            className="label px-0 pb-1"
          >
            <span
              className="label-text font-bold text-white"
            >
              Email
            </span>
          </div>
          <input
            className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type here"
            type="text"
            value=""
          />
        </label>
        <label
          className="lg:min-w-xs form-control w-full"
        >
          <div
            className="label px-0 pb-1"
          >
            <span
              className="label-text font-bold text-white"
            >
              Date of birth
            </span>
          </div>
          <input
            className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
            name="dateOfBirth"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type here"
            type="date"
            value=""
          />
        </label>
        <label
          className="lg:min-w-xs form-control w-full"
        >
          <div
            className="label px-0 pb-1"
          >
            <span
              className="label-text font-bold text-white"
            >
              Address
            </span>
          </div>
          <input
            className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
            name="address"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type here"
            type="text"
            value=""
          />
        </label>
        <label
          className="lg:min-w-xs form-control w-full"
        >
          <div
            className="label px-0 pb-1"
          >
            <span
              className="label-text font-bold text-white"
            >
              City
            </span>
          </div>
          <input
            className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
            name="city"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Type here"
            type="text"
            value=""
          />
        </label>
        <div
          className="flex flex-col gap-4 lg:w-full lg:flex-row"
        >
          <label
            className="lg:min-w-xs form-control w-full"
          >
            <div
              className="label px-0 pb-1"
            >
              <span
                className="label-text font-bold text-white"
              >
                State
              </span>
            </div>
            <input
              className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
              name="state"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Type here"
              type="text"
              value=""
            />
          </label>
          <label
            className="lg:min-w-xs form-control w-full"
          >
            <div
              className="label px-0 pb-1"
            >
              <span
                className="label-text font-bold text-white"
              >
                Zip code
              </span>
            </div>
            <input
              className="lg:min-w-xs input input-bordered w-full bg-white text-info-content placeholder:text-gray-400"
              name="zipCode"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Type here"
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col gap-8 rounded-3xl bg-white p-6 lg:justify-between lg:gap-4"
    >
      <div
        className="flex flex-col gap-8"
      >
        <h2
          className="text-2xl font-bold uppercase text-info-content"
        >
          Summary
        </h2>
        <div
          className="flex flex-col items-center gap-2"
        >
          <div
            className="flex size-48 items-center justify-center rounded-full bg-gray-100"
          >
            <svg
              className="size-24"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.5 7h-.7c-.4-1.2-1.5-2-2.8-2s-2.4.9-2.8 2.1c-.3-.4-.7-.6-1.2-.6s-.9.2-1.2.6C6.4 5.9 5.3 5 4 5s-2.4.9-2.8 2H.5c-.3 0-.5.2-.5.5s.2.5.5.5H1c0 1.7 1.3 3 3 3c1.5 0 2.7-1.1 3-2.5c.3 0 .5-.2.5-.5s.2-.5.5-.5s.5.2.5.5s.2.5.5.5c.2 1.4 1.5 2.5 3 2.5c1.7 0 3-1.3 3-3h.5c.3 0 .5-.2.5-.5s-.2-.5-.5-.5M4 10c-1.1 0-2-.9-2-2s.9-2 2-2s2 .9 2 2s-.9 2-2 2m8 0c-1.1 0-2-.9-2-2s.9-2 2-2s2 .9 2 2s-.9 2-2 2"
                fill="currentColor"
              />
            </svg>
          </div>
          <span
            className="text-center text-base font-bold text-info-content"
          >
            Minifig
          </span>
        </div>
        <div
          className="flex flex-col gap-4"
        >
          <span
            className="font-bold text-info-content"
          >
            There are 
            2
             parts in this minifig:
          </span>
          <div
            className="flex gap-3"
          >
            <div
              className="flex flex-col"
            >
              <span
                className="line-clamp-1 text-sm font-bold text-info-content"
              >
                Part 1
              </span>
              <span
                className="text-xs text-accent"
              >
                p1
              </span>
            </div>
          </div>
          <div
            className="flex gap-3"
          >
            <div
              className="flex flex-col"
            >
              <span
                className="line-clamp-1 text-sm font-bold text-info-content"
              >
                Part 2
              </span>
              <span
                className="text-xs text-accent"
              >
                p2
              </span>
            </div>
          </div>
        </div>
      </div>
      <button
        className="btn btn-primary w-full disabled:text-info-content"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`ChooseMinifig > should create a snapshot with empty cart 1`] = `
<div
  className="flex min-h-dvh items-center justify-center p-4 xl:mx-auto xl:w-maxWidth"
>
  <div
    className="flex w-full flex-col items-center gap-12 rounded-3xl bg-white p-4 lg:w-fit lg:p-12"
  >
    <div
      className="flex flex-col gap-2"
    >
      <h1
        className="text-center text-2xl font-bold uppercase text-info-content"
      >
        Your cart is empty!
      </h1>
      <p
        className="text-center text-base text-gray-400"
      >
        Please choose your minifig.
      </p>
    </div>
    <a
      className="w-full lg:w-fit"
      href="/choose-minifig"
      onClick={[Function]}
    >
      <button
        className="btn btn-primary w-full text-white"
      >
        Select minifig
      </button>
    </a>
  </div>
</div>
`;
